### ISSUE TITLE
[Feature Request] Filter Features Input Transform for Deterministic Models

### ISSUE THREAD

**Issue by jduerholt** [pytorch/botorch#1511]
# 🚀 Feature Request

Hi, what are you thinking about adding something like the `FilterFeatures` `InputTransform` for deterministic models?

## Motivation

We have sometimes custom deterministic models that implement something like an empirical cost formula for a material etc.. We would like to store these models and reuse them in different optimization scenarios. Imagine that the cost function acts on features A, B and C. In one optimization we optimize in addition also over features D and E and in another one only about A,B, C and D (the features also enter GP models for the other desired outputs). To reuse the custom deterministic model it would be the best to program it in a way that it only acts on three features and depending on the situation setup something like a Features filter that filters out the indices of the features that are not needed for the deterministic model. One can then just instantiate the model (or load from pickle), set the filter and use it in the new optimization. The `subset_output` method in `AffineDeterministicModel` goes already in this direction.

It could be also of interest to bring the whole logic of input transforms also to deterministic models, but this would be a bigger effort.

## Pitch

**Describe the solution you'd like**
For the start, I would propose to add an addtional attribute to `DeterministicModel` like `indices` and filter with those indices in the `forward` function before actually executing the real forward. Derived models would then in their `forward` method first call the `forward` method of the parent function and then their own custom model implementation.  

**Describe alternatives you've considered**
Do not build it directly into botorch but into the custom deterministic model classes when setting them up. But I think it would be nice if this would be supported directly by botorch. 

**Are you willing to open a pull request?** (See [CONTRIBUTING](../CONTRIBUTING.md))
Yes.

**Comment by Balandat** [comment]
Thanks for the suggestion!

> It could be also of interest to bring the whole logic of input transforms also to deterministic models, but this would be a bigger effort.

I think it would make a lot of sense to enable input transforms more generally for deterministic models, rather than adding some idiosyncratic functionality for subsetting. I agree that it'll be a bit more work but it shouldn't really be all that bad. Then nice thing is that since there isn't any training data whose transforms need to be handled under the hood (which is the main complication for stateful models), this should be rather straightforward.

Basically we'll just have the `DeterminsiticModel`  base class constructor take an optional `InputTransform`, assign that to the instance, and then call it in the `forward` method. Then the other derived models can call that. I think the main question is how we can make this be applied automatically in a way that doesn't require every user to add some additional code for transforming in their `forward` methods. @saitcakmak will probably have some good thoughts here.

**Comment by saitcakmak** [comment]
I actually implemented this a while ago. You just need to set `det_model.input_transform = ...` and call it through `det_model.posterior(...)`, just like you'd do with any other model. It supports the outcome transforms as well. Full posterior implementation is [here](https://github.com/pytorch/botorch/blob/main/botorch/models/deterministic.py#L65-L108).

**Comment by saitcakmak** [comment]
> Basically we'll just have the DeterminsiticModel base class constructor take an optional InputTransform, assign that to the instance, and then call it in the forward method. 

The base `DeterminsiticModel` does not have a constructor, so the current implementation leaves it to each class to independently do the assignment. The `forward` method also doesn't use the transform (since this would have to be for each class) but that shouldn't really be an issue since we do no train `DeterministicModel`s, so no need to call `forward` rather than `posterior` ever. (This will likely change once I get back to the input transform refactor, calling `det_model(..)` might just work after that). 

> I think the main question is how we can make this be applied automatically in a way that doesn't require every user to add some additional code for transforming in their forward methods.

It has been a while since I worked on it, but this might be handled as part of the input transform refactor. Implementing a `__call__` for `DeterministicModel` and applying it there the same way we (will) do it in `ExactGP` would be an option.

**Comment by jduerholt** [comment]
Thanks for the fast feedback. I just tested it, it worked like a charm! I really like input transforms in botorch!
